### DISCUSSION TITLE
client.query, client.mutation returning rawResponse for toPromise

### DISCUSSION THREAD

**Discussion by emmyakin** [urql-graphql/urql#2452] [Q&A]
Is it feasible for the client.toPromise API to return the raw fetch Response 

i.e `OperationResult` to include `rawFetchResponse`

In order to access headers, it is required to pass in a custom fetch implementation, however, when testing using steps outlined in https://formidable.com/open-source/urql/docs/advanced/testing/, the custom fetch implementation is not called, and so, assertions cannot be made against logic used in the custom fetch implementation, for example, where the headers is accessed and returned alongside the data

**Comment by emmyakin** [comment]
I am reading about exchanges a bit more, looks like I may find an answer there

**Comment by kitten** [comment]
That's not possible, no — but for a very specific reason.
Generally speaking, we don't impose many opinions, but the opinion we do consistently apply is around transport protocols. Most commonly, GraphQL is executed via HTTP(s), so via HTTP. But that means we have GraphQL _wrapped_ in HTTP. So, how we see it, and how the [GraphQL over HTTP spec](https://github.com/graphql/graphql-over-http) implies, you're either getting a raw-HTTP response in error scenarios, or you're getting a GraphQL response.

That means that typically a GraphQL response is fully enclosed and should refrain from returning additional metadata in anything but the specc'ed response. That's why having an `extensions` key on the GraphQL JSON response is common, or why GraphQL specs its errors in JSON results to also have `extensions` fields. This is to deliver untyped metadata about the response, i.e. protocol extensions.
This has the benefit of GraphQL remaining protocol-agnostic. This means that you will always be able to switch to, say, GraphQL over WebSockets and still get the same results, and potentially the same metadata.

A common exception is authentication. There we often get to choose how we handle expired auth tokens, or expired sessions. There we make an assumption that authentication errors will still be delivered in the GraphQL JSON format, i.e. with auth errors as `{ errors: [{ message: '...', extensions: { code: 'EXPIRED_SESSION' }}] }` for instance.
https://formidable.com/open-source/urql/docs/advanced/authentication/

But generally speaking, when a response doesn't adhere to the GraphQL result format, we throw it away and assume it's an error in our default `fetchExchange`.

**Comment by emmyakin** [comment]
Thank you. This is enlightening.
